tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/DistanceTable.java

```diff
@@ -0,0 +1,556 @@
+        // subtable information.
+        private void writeDistancePlusSubtables(Span trie) {
+            trie.putPrefixAndValue(distance);
+            writeTo(trie);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder buffer = new StringBuilder("distance: ").append(distance).append('\n');
+            return appendToString("", buffer).toString();
+        }
+
+        private StringBuilder appendToString(String indent, StringBuilder buffer) {
+            // Top level values are not padded with tabs.
+            String rowIndent = indent.isEmpty() ? "" : "\t";
+            for (Map.Entry<String, Map<String, Node>> row : subtables.rowMap().entrySet()) {
+                buffer.append(rowIndent).append(row.getKey());
+                // First column extends the current row, so single tab indent.
+                String colIndent = "\t";
+                for (Map.Entry<String, Node> col : row.getValue().entrySet()) {
+                    buffer.append(colIndent).append(col.getKey());
+                    Node subnode = col.getValue();
+                    buffer.append('\t').append(subnode.distance);
+                    // Append any sub-nodes (starting on the same line).
+                    subnode.appendToString(indent + "\t\t\t", buffer).append('\n');
+                    // Later columns need full indent (including skipping row key).
+                    colIndent = indent + '\t';
+                }
+                // Later rows need full indent.
+                rowIndent = indent;
+            }
+            return buffer;
+        }
+    }
+
+    /**
+     * Excapsulates a sequence of {@code <desired,supported>} pairwise mappings over
+     * language, script and region, with an associated distance. This is an alternate
+     * way to represent a mapping of desired and supported language match rules.
+     *
+     * <p>For example:
+     * <pre>{@code
+     *   <languageMatch desired="en_*_$!enUS", supported="en_*_$GB", distance="3"/>
+     * }</pre>
+     * results in a set of keys of the form:
+     * <pre>{@code
+     *   <en,en> -> <ANY,ANY> -> <X,Y> = 3
+     * }</pre>
+     * where the "region" part {@code <X,Y>} is constructed from all the possible
+     * combinations of partition IDs associated with the original region variables.
+     *
+     * <p>Mapping keys have several useful properties:
+     * <ul>
+     *     <li>They can be reversed (e.g. {@code <A,B> -> <C,D> = N} becomes
+     *     {@code <B,A> -> <D,C> = N}).
+     *     <li>They can be symmetrical (e.g. {@code <X,X> -> <Y,Y> = N}), in which
+     *     case the reversed key is the same as the original.
+     *     <li>They can have wildcard mappings (i.e. {@code <ANY,ANY>}).
+     *     <li>They can produce "suffix" keys (e.g. the suffix of
+     *     {@code <A,B> -> <C,D> = N} is {@code <C,D> = N}).
+     * </ul>
+     */
+    private static final class MappingKey {
```

Comment:
While this is a fair number of lines of code, having an encapsulation of the key pairs allows all levels of the table to be handled with the same code, and makes actual business logic shorter and more readable.